### DISCUSSION TITLE
Inconsistencies in file sizes and modification times

### DISCUSSION THREAD

**Discussion by xyb** [juicedata/juicefs#3777] [General]
**What happened**:
When listing the files in a directory mounted from remote JuiceFS (MySQL + Aliyun OSS), I was surprised to discover that some files had zero sizes and inconsistent modification times across different hosts. Here are the observed differences:
```diff
--- list.a	2023-06-05 16:56:31.000000000 +0800
+++ list.b	2023-06-05 16:57:55.000000000 +0800
@@ -1,7 +1,7 @@
-# Host A, with correct stats
+# Host B
 $ TZ=utc ls -l
 total 730381
--rw-r--r-- 1 root root    331177 May 30 05:42 'D & I.doc'
+-rw-r--r-- 1 root root         0 May 30 05:40 'D & I.doc'
 -rw-r--r-- 1 root root     93411 May 30 05:46  T.doc
 -rw-r--r-- 1 root root    356212 May 30 05:46  T.zip
 -rw-r--r-- 1 root root   4296509 May 30 05:28  不.doc
@@ -9,11 +9,11 @@
 -rw-r--r-- 1 root root      4281 May 30 15:51  专.xlsx
 -rw-r--r-- 1 root root     33074 May 30 15:51 '专.doc'
 -rw-r--r-- 1 root root     32827 May 30 15:51  专-.doc
--rw-r--r-- 1 root root   4130579 May 30 05:28  昨.doc
+-rw-r--r-- 1 root root   4130579 May 30 05:22  昨.doc
 -rw-r--r-- 1 root root    261104 May 30 15:51 '歌.doc'
 -rw-r--r-- 1 root root     61164 May 30 15:51  歌-.doc
 -rw-r--r-- 1 root root 522505700 May 30 15:53 '温.zip'
 -rw-r--r-- 1 root root 213707138 May 30 15:53  石.zip
--rw-r--r-- 1 root root   1635274 May 30 05:45  石.doc
--rw-r--r-- 1 root root     97784 May 30 05:47  高.doc
+-rw-r--r-- 1 root root   1635274 May 30 05:27  石.doc
+-rw-r--r-- 1 root root         0 May 30 05:47  高.doc
 -rw-r--r-- 1 root root    355296 May 30 05:47  高.zip
```

Furthermore, I also have a Docker container, which has mounted the same directory from Host B as a volume, is reporting an incorrect file size, although it is not zero:
```diff
--- list.a	2023-06-05 17:08:22.000000000 +0800
+++ list.c	2023-06-05 17:08:34.000000000 +0800
@@ -1,7 +1,7 @@
-# Host A, with correct stats
+# Container C
 $ TZ=utc ls -l
 total 730381
--rw-r--r-- 1 root root    331177 May 30 05:42 'D & I.doc'
+-rw-r--r-- 1 root root    331177 May 30 05:41 'D & I.doc'
 -rw-r--r-- 1 root root     93411 May 30 05:46  T.doc
 -rw-r--r-- 1 root root    356212 May 30 05:46  T.zip
 -rw-r--r-- 1 root root   4296509 May 30 05:28  不.doc
@@ -9,11 +9,11 @@
 -rw-r--r-- 1 root root      4281 May 30 15:51  专.xlsx
 -rw-r--r-- 1 root root     33074 May 30 15:51 '专.doc'
 -rw-r--r-- 1 root root     32827 May 30 15:51  专-.doc
--rw-r--r-- 1 root root   4130579 May 30 05:28  昨.doc
+-rw-r--r-- 1 root root   4130579 May 30 05:27  昨.doc
 -rw-r--r-- 1 root root    261104 May 30 15:51 '歌.doc'
 -rw-r--r-- 1 root root     61164 May 30 15:51  歌-.doc
--rw-r--r-- 1 root root 522505700 May 30 15:53 '温.zip'
--rw-r--r-- 1 root root 213707138 May 30 15:53  石.zip
+-rw-r--r-- 1 root root 363327488 May 30 15:52 '温.zip'
+-rw-r--r-- 1 root root         0 May 30 15:51  石.zip
 -rw-r--r-- 1 root root   1635274 May 30 05:45  石.doc
 -rw-r--r-- 1 root root     97784 May 30 05:47  高.doc
 -rw-r--r-- 1 root root    355296 May 30 05:47  高.zip
```
After recreating container C, it now displays the same content as its host B.

**What you expected to happen**:
Consistencies in file stats

**How to reproduce it (as minimally and precisely as possible)**:
Don't know how to reproduct it right now.

**Anything else we need to know?**

**Environment**:
- JuiceFS version (use `juicefs --version`) or Hadoop Java SDK version: 1.0.4+2023-04-06.f1c475d
- Cloud provider or hardware configuration running JuiceFS: Aliyun ECS
- OS (e.g `cat /etc/os-release`): Ubuntu 22.04.1 LTS
- Kernel (e.g. `uname -a`): 5.15.0-53-generic
- Object storage (cloud provider and region, or self maintained): oss-cn-beijing.aliyuncs.com
- Metadata engine info (version, cloud provider managed or self maintained): Aliyun MySQL 8.0
- Network connectivity (JuiceFS to metadata engine, JuiceFS to object storage): internal
- Others:

**Comment by xyb** [comment]
Update:
When accessing a freshly mounted directory, it consistently displays correct file stats.

**Comment by SandyXSD** [comment]
What's the mount option or container configuration for juicefs? Are these files being updated?
Usually the inconsistency is caused by kernel or process metadata cache.

**Comment by xyb** [comment]
> What's the mount option or container configuration for juicefs?

```sh
❯ grep Option /etc/systemd/system/multi-user.target.wants/juicefs.mount
Options=_netdev,allow_other,writeback_cache
```

> Are these files being updated?

Sometimes my application needs to recreate entire files dynamically.

If this can potentially cause issues, I would like to know how I can prevent any interference from the kernel or process metadata cache.

**Comment by SandyXSD** [comment]
Try removing the `writeback_cache`

**Reply by xyb** [reply]
After removing the `writeback_cache` option, the same issue has not reoccurred after monitoring for a week. The problem has been resolved, thank you!

**Comment by davies** [comment]
JuiceFS guarantee close-to-open, which means another client will see the changes (newly written data, or size, mtime) AFTER the file is closed. When you saw different size and mtime on different clients, which means the client with newer mtime has not close the file yet.

So even without writeback_cache, you may still the above behavior.

Within same client (same mount point), it guarantee read-after-write, which means the size and mtime is always up-to-date.

**Comment by solracsf** [comment]
@davies there is no mention of open-after-close [on docs](https://juicefs.com/docs/community/introduction/). Only close-to-open is mentioned.

![image](https://github.com/juicedata/juicefs/assets/12234510/fc1e53a5-3bf6-4afb-aff6-1baf1abe864b)

**Comment by davies** [comment]
They are different words on the same thing. `close-to-open` is the correct name, I have update the comment.
